### ISSUE TITLE
useShape sync broken on Expo react native - CORS config error and missing headers warning

### ISSUE THREAD

**Issue by federico-hv** [electric-sql/electric#2072]
Hi, first of all thanks for this amazing piece of tech. I love Postgres and I'm really excited about the future of Electric. I was recently trying the new version and I deployed the service to a Digital Ocean Droplet with Docker as recommended on the docs https://electric-sql.com/docs/integrations/digital-ocean. Worth noting I had to set up an IPV6 network to connect from within to Supabase (maybe this could be a good addition to the docs). Now, the service seems to be working fine on the initial sync and I get all the headers right while testing with curl but I'm having one issue with the `useShape` throwing an warning on react-native when I make updates to the database and not retrieving the latest data (only react-native on web seems to work fine) . Also, I know the hook is only for read-only (that's what I'm testing it for while making updates directly to the DB). I'm attaching a screenshot of the error, it seems that there are missing headers but this is only happening on react native (Expo) not on web. I checked the service with curl but it already has the headers and seems ok, however I read some people are using services on top (proxies) like hono.js and they had to use a specific [CORS config](https://github.com/electric-sql/electric/issues/1951) for these headers to be preserved. I tried adding an express proxy on top but the issue is still there. Any ideas on what could be causing this issue? Just wondering if I missed any configuration? 

![Screenshot 2024-11-29 at 2 47 27 PM](https://github.com/user-attachments/assets/8e943467-6851-4a44-b415-db6c33310e00)

<img width="969" alt="Screenshot 2024-11-29 at 3 31 31 PM" src="https://github.com/user-attachments/assets/8b7bc8ef-5a72-4533-90ee-76d8d2102981">

**Comment by KyleAMathews** [comment]
Hmm so the schema isn't returned on live quests (when request includes `?live`) — so this error shouldn't happen as the client only looks for `electric-cursor` not `electric-schema` — https://github.com/electric-sql/electric/blob/32d4b5085a3c54a53216e37ad230a2fb79fdbb1f/packages/typescript-client/src/fetch.ts#L184-L189

Do live updates come through? I'm not very familiar with why Expo wouldn't be working here. Perhaps the fetch implementation is different somehow?

**Comment by federico-hv** [comment]
Hey @KyleAMathews, thanks for checking this. So after digging a bit, it seems the issue was not on the electric typescript client but on the implementation of the URL api that Expo uses. The method `url.searchParams.has` matches true even when passing an incorrect value to match as second argument.  I created an issue to explain this on the `whatwg-url-without-unicode` repo https://github.com/charpeni/whatwg-url/issues/7 and another one on the expo repo to mention the problem https://github.com/expo/expo/issues/33336. In the meantime the solution to the client breaking in react-native would be to avoid the calls that use the second argument or just use another package optimized for URLS like https://github.com/sindresorhus/query-string or validate the string directly with something like `includes`,  `indexOf`, `match`, `test`, etc -> Although this may affect performance. 

```js

// Replacing these would fix the issue

url.searchParams.has(LIVE_QUERY_PARAM, `true`)
url.searchParams.has(LIVE_QUERY_PARAM, `false`) // returns true even when live=true
```


Let me know your thoughts on this. Happy to add a fix, test it and create a PR.

**Comment by KyleAMathews** [comment]
Huh, not common to find a bug in a url lib so kudos!

Yeah a PR just replacing the has call with a get and value check would work great.

**Comment by federico-hv** [comment]
Hey @KyleAMathews, sorry I didn't end up creating the PR to close this one on time. I actually was trying to understand better what was going on with the URL stuff in the react-native ecosystem and after some testing I thought the best way to provide compatibility is to not expect any Web APIs to be fully functional for now.  

Just for reference I did some research on url parsing libs and urlite seems to be generic, [fast and small](https://www.npmjs.com/package/urlite#comparison) so I made this #2116 PR that removes the use of the URL object completely in favor of the generic parser in case you think it can be an improvement, otherwise feel free to discard it. 


Anyways, your recent change fixed the issue so thanks!!! Now the typescript-client seems stable in Expo.

 
![screen_gif](https://github.com/user-attachments/assets/bd9cd8a4-c5ea-424b-a041-3cacac438bb0)
